### DISCUSSION TITLE
Questions about "OslcDialog" in Service Provider

### DISCUSSION THREAD

**Discussion by lll4770** [OSLC/oslc4net#83] [Q&A]
Hi,
I'm a new beginner of OSLC, and I'm learning on your project.
I studied the StockQuoteSample project, I noticed we could add the "OslcDialog" and "OslcQueryCapability" on the same function(GetStockQuotes). From my understanding, this function "GetStockQuotes" is a "GET" function, why can you put the annotation "OslcDialog" on this function? In other words, what functions can we declare as "OslcDialog", and how about the "url" of "OslcDialog"?

**Comment by Jad-el-khoury** [comment]
Hi 

May I suggest you ask these questions on the OSLC forum https://forum.open-services.net/ ? 
This is a general OSLC enquire and I am sure the broarder OSLC community can better help you there.
Jad

**Comment by lll4770** [comment]
Got it, thanks @Jad-elkhouy

**Comment by Jad-el-khoury** [comment]
And please add links to the piece of code you are referring to. So we don't have to guess what you mean.

**Comment by lll4770** [comment]
I'm going to sign up on the OSLC forum... 
And what I want to know is this https://github.com/OSLC/oslc4net/blob/master/OSLC4Net_SDK/OSLC4Net.StockQuoteSample/Controllers/StockQuoteController.cs#L53; I noticed that the "OslcDialog" and "OslcQueryCapability" are declared on the same function (GetStockQuotes), I also want to know how OslcDialog works

**Comment by jamsden** [comment]
Class StockQuoteController provides the OSLC server for stock quotes. That server provides OSLC delegated dialog, creation factory and query capabilities on stock quotes. Function getStockQuote is a method that returns an RDF representation of a stock quote, it is not directly related to delegated dialogs, creation factories or query capabilities. Rather it is just part of the POST, GET, PUT, DELETE HTTP methods on stock quotes that a typical OSLC server would provide. 

I suggest closing this issue and moving the discussion to the OSLC forum.

**Comment by berezovskyi** [comment]
@lll4770 welcome to OSLC4Net! You are welcome to post questions here or on the forum. But please use https://github.com/OSLC/oslc4net/discussions in the future for the questions, if you prefer to stay on Github. ~~But we can finish discussing this question here, such as not to fragment the conversation.~~ I found that GH has a feature to move conversations to Discussions.

Regarding your questions:

> [...] function "GetStockQuotes" is a "GET" function, why can you put the annotation "OslcDialog" on this function?

According to the OSLC Core spec, both OSLC Query Capability and the OSLC Selection Dialog endpoints should be requested via an HTTP GET request. There are now discussions in the OSLC Open Project to change this (which I don't support much, as I believe that all idempotent requests should be made using idempotent HTTP verbs, but that's another topic). So, there is no problem with the fact that the method is only designed to handle GET requests.

> I noticed we could add the "OslcDialog" and "OslcQueryCapability" on the same function(GetStockQuotes). [...] why can you put the annotation "OslcDialog" on this function? In other words, what functions can we declare as "OslcDialog", and how about the "url" of "OslcDialog"?
 
I specifically checked the code behavior and can say that for the time being, these attributes (annotations are called [attributes](https://learn.microsoft.com/en-us/dotnet/csharp/advanced-topics/reflection-and-attributes/) in C#) are only used to build Service and Service Provider definitions (e.g., when you request `curl --location 'http://localhost:7077/api/serviceprovider' --header 'Accept: application/rdf+xml'`). This means the attribute will not have any effect on the routes definition in the `WebApiConfig` and there will be no controller action to invoke (more on the [controller and action selection algorithms](https://learn.microsoft.com/en-us/aspnet/web-api/overview/web-api-routing-and-actions/routing-and-action-selection#selecting-a-controller) in .NET MVC). Adding such support will be tricky given that the standard `ActionName` attribute has [`AllowMultiple=false`](https://learn.microsoft.com/en-us/dotnet/api/system.web.mvc.actionnameattribute?view=aspnet-mvc-5.2), which does not allow to easily assign multiple controller actions to the same method.

Another issue is that you will need to return RDF from a query endpoint (but could be HTML if you want to provide a more rich experience), while the Selection Dialog is HTML-only. Thus, you will need to juggle between returning an `ICollection<IResource>` or (likely) Razor templates.

All said, I suggest you define a common method (non-public and/or annotated with the [`[NonAction]` attribute](https://learn.microsoft.com/en-us/aspnet/web-api/overview/web-api-routing-and-actions/routing-in-aspnet-web-api#non-actions)) and create two other methods that could be used as controller actions for query and selection, each annotated with a corresponding attribute.

Finally, thank you for bringing this up, we will add an issue to update the sample code to avoid this use of the attributes that are not backed by a matching controller action. But I think it will remain in the backlog until we develop a new sample, given that this sample uses a stock exchange information model that is quite far from the OSLC application area at the moment (engineering toolchain integration).

**Comment by berezovskyi** [comment]
> I also want to know how OslcDialog works

For that, I would like to recommend using the Java reference implementation to see how the dialogs work: https://github.com/oslc-op/refimpl/

Technically, they use HTML iframes and the [postMessage Javascript protocol](https://developer.mozilla.org/en-US/docs/Web/API/Window/postMessage) to exchange the data between the client page and the page (e.g., a selection dialog) embedded in an iframe.

**Reply by lll4770** [reply]
Hi @berezovskyi , thanks so much for your detailed explanation, it really helped me understand it better.
I read some articles about OSLC and found that this project wouldn't be maintained. I'm going to move on the Java implementation (just like you recommended), if I have any questions I will post it on your Forum. Thanks again!

**Reply by berezovskyi** [reply]
Glad to help! As you see, I made the [first release on Nuget](https://github.com/OSLC/oslc4net/discussions/73) recently, so the project is no longer unmainained. But we still need more contributors and cleanup to recommend OSLC4Net for use in production. Eclipse Lyo, on the other hand is used by multiple companies in production. By the way, may I ask you if need Net4.8 support or would you be ok with OSLC4Net targeting NET Core only?

**Reply by lll4770** [reply]
Oh, that's great! Both .Net4.8 and .Net Core make sense to me.
To be honest, my company has two PLM products, one is developed by .NET and the other is developed by Java, and we have the plan to implement OSLC in the both PLM products, so I'm exploring how OSLC works and I'd like to run a sample in my local environment to see the whole progress of OLSC which will be of great help for us in our future development.

**Comment by Jad-el-khoury** [comment]
@berezovskyi ! Why are we introducing Discussions under Github? I would have thought it better to consolidate all discussions under our Forum.

**Reply by berezovskyi** [reply]
For the OSLC4Net project specifically, I decided to enable Discussions to see if keeping everying in one place will help.

**Reply by Jad-el-khoury** [reply]
Yes, keeping everying in one place will help.
Now discussions are in two places. Besides Jim, you and I, I doubt anyone else will take note of what is happening here.
